src/main/java/com/squareup/javawriter/builders/TypeSpec.java

```diff
@@ -42,6 +43,17 @@
   private TypeSpec(Builder builder) {
     checkArgument(builder.name != null ^ builder.anonymousTypeArguments != null,
         "types must have either a name or anonymous type arguments");
+    boolean isInterface = builder.type == Type.INTERFACE;
+    boolean typeIsAbstract = builder.modifiers.contains(Modifier.ABSTRACT) || isInterface;
+    for (MethodSpec methodSpec : builder.methodSpecs) {
+      checkArgument(typeIsAbstract || !methodSpec.hasModifier(Modifier.ABSTRACT),
+          "non-abstract type %s cannot declare abstract method %s", builder.name, methodSpec.name);
+      checkArgument(!isInterface || methodSpec.hasModifier(Modifier.ABSTRACT),
+          "interface %s cannot declare non-abstract method %s", builder.name, methodSpec.name);
+      checkArgument(!isInterface || methodSpec.hasModifier(Modifier.PUBLIC),
```

Comment:
We'll have to go through and add tests which verify these conditions correctly fail.